pinecone/core/grpc/index_grpc.py

```diff
@@ -299,12 +311,37 @@ def _vector_transform(item):
 
         timeout = kwargs.pop('timeout', None)
 
-        request = UpsertRequest(vectors=list(map(_vector_transform, vectors)), **args_dict, **kwargs)
+        vectors = list(map(_vector_transform, vectors))
         if async_req:
+            args_dict = self._parse_non_empty_args([('namespace', namespace)])
+            request = UpsertRequest(vectors=vectors, **args_dict, **kwargs)
             future = self._wrap_grpc_call(self.stub.Upsert.future, request, timeout=timeout)
             return PineconeGrpcFuture(future)
-        else:
-            return self._wrap_grpc_call(self.stub.Upsert, request, timeout=timeout)
+
+        if batch_size is None:
+            return self._upsert_batch(vectors, namespace, timeout=timeout, **kwargs)
+
+        if not isinstance(batch_size, int) or batch_size <= 0:
+            raise ValueError('batch_size must be a positive integer')
+
+        pbar = tqdm(total=len(vectors), disable=not show_progress, desc='Upserted vectors')
+        total_upserted = 0
+        for i in range(0, len(vectors), batch_size):
+            batch_result = self._upsert_batch(vectors[i:i + batch_size], namespace, timeout=timeout, **kwargs)
+            pbar.update(batch_result.upserted_count)
+            # we can't use here pbar.n for the case show_progress=False
+            total_upserted += batch_result.upserted_count
+
+        return UpsertResponse(upserted_count=total_upserted)
+
+    def _upsert_batch(self,
+                      vectors: List[GRPCVector],
+                      namespace: Optional[str],
+                      timeout: Optional[float],
+                      **kwargs) -> UpsertResponse:
+        args_dict = self._parse_non_empty_args([('namespace', namespace)])
+        request = UpsertRequest(vectors=vectors, **args_dict)
+        return self._wrap_grpc_call(self.stub.Upsert, request, timeout=timeout, **kwargs)
```

Comment:
directly return the batch_result.upserted_count will clean up the code on the calling function

Reply:
It's a matter of taste, but I prefer it this way for several reasons:

1. This way we can return this result directly when `batch_size` is None, rather then re-create a new response object for no reason
2. Usually when I write functions that do IO I like to have their return type to be the IO result to keep it clean and doing one thing as possible
3. My idea about writing the handle in `batch_size` like this is to have the same code running as we had before, to take as least risk as possible for breaking changes, and here I believe that in the near future 99% of the users will not use the batch size yet

Reply:
it makes sesnse for now, but generally, the lower level functions should handle result parsing